working/0110-incremental-sound-nnbd/feature-specification.md

```diff
@@ -0,0 +1,374 @@
+`Null` in the `false` continuation.
+
+The static type of an expression `e!` is **NonNull**(`T`) where `T` is the
+static type of `e`.
+
+The **NonNull** function defines the null-promoted version of a type, and is
+defined as follows.
+
+- **NonNull**(Null) = Never
+- **NonNull**(_C_<_T_<sub>1</sub>, ... , _T_<sub>_n_</sub>>) = _C_<_T_<sub>1</sub>, ... , _T_<sub>_n_</sub>>  for class *C* other than Null
+- **NonNull**(FutureOr<_T_>) = FutureOr<_T_>   
+- **NonNull**(_T_<sub>0</sub> Function(...)) = _T_<sub>0</sub> Function(...)
+- **NonNull**(Function) = Function
+- **NonNull**(Object) = Object
+- **NonNull**(Never) = Never
+- **NonNull**(dynamic) = dynamic
+- **NonNull**(void) = void   
+- **NonNull**(_X_) = X & **NonNull**(B), where B is the bound of X.
+- **NonNull**(_X_ & T) = X & **NonNull**(T)
+- **NonNull**(_T_?) = **NonNull**(_T_)
+- **NonNull**(_T_\*) = **NonNull**(_T_)
+
+#### Type promotion, Definite Assignment, and Definite Completion
+
+These are extended as per separate proposal.
+
+### Runtime semantics
+
+#### Null assertion operator
+
+An expression of the form `e!` evaluates `e` to a value `v`, throws a runtime
+error if `v` is `null`, and otherwise evaluates to `v`.
+
+#### Null aware operator
+
+An expression of the form `e?.<tail>` where `<tail>` is any sequence of
+selectors evaluates to `null` if `e` evaluates to `null`, and otherwise
+evaluates to the same value as `e.<tail>`.
+
+An expression of the form `e?..<tail>` where `<tail>` is any sequence of
+selectors evaluates to `null` if `e` evaluates to `null`, and otherwise
+evaluates to the same value as `e..<tail>`.
+
+**TODO** Define exactly how a valid `<tail>` is delimited.
+
+#### Lazy fields and variables
+
+A read of a field or variable which is marked as `lazy` which has not yet been
+written to causes the initializer expression of the variable to be evaluated to
+a value, assigned to the variable or field, and returned as the value of the
+read.
+  - If there is no initializer expression, the read causes a runtime error.
+  - Evaluating the initializer expression may validly cause a write to the field
+    or variable, assuming that the field or variable is not final.  In this
+    case, the variable assumes the written value.  The final value of the
+    initializer expression overwrites any intermediate written values.
+  - Evaluating the initializer expression may cause an exception to be thrown.
+    If the variable was written to before the exception was thrown, the value of
+    the variable on subsequent reads is the last written value.  If the variable
+    was not written before the exception was thrown, then the next read attempts
+    to evaluate the initializer expression again.
+  - If a variable or field is read from during the process of evaluating its own
+    initializer expression, and no write to the variable has occurred, the read
+    is treated as a first read and the initializer expression is evaluated
+    again.
```

Comment:
I don't think we should recurse here. If a lazy variable is read during the evaluation of its own initializer expression, the read should throw (a `CyclicInitiaizationError`) like it currently does.

Reply:
We had a long discussion of this on the board, and I thought we had resolved this in favor of not doing the cyclic initialization error checks.  Have you changed your mind on this?

Reply:
Ack, no, just didn't remember it. My first instinct is still to disallow what is almost certainly an infinite recursion or an early throw (a second attempt to store an initializer expression's value for a final field will throw). 
I can see the arguments for not doing extra state-management (with a need for try/finally), but it just feels so very fragile.

But, yes, this is what we agreed on.